src/Diagnostics/average.jl

```diff
@@ -0,0 +1,85 @@
+using Oceananigans.Architectures
+using Oceananigans.BoundaryConditions
+using Oceananigans.Utils
+using Oceananigans.Grids: total_size
+
+"""
+    Average{F, R, D, P, I, T} <: AbstractDiagnostic
+
+A diagnostic for computing the averages of a field along particular dimensions.
+"""
+mutable struct Average{F, R, D, P, I, T} <: AbstractDiagnostic
+                 field :: F
+                  dims :: D
+                result :: P
+    iteration_interval :: I
+         time_interval :: T
+              previous :: Float64
+           return_type :: R
+end
+
+function dims_to_result_size(field, dims, grid)
+    N = (grid.Nx, grid.Ny, grid.Nz)
+    field_size = total_size(parent(field))
+    return Tuple(d in dims ? 1 : field_size[d] for d in 1:3)
+end
+
+"""
+    Average(field; dims, iteration_interval=nothing, time_interval=nothing, return_type=Array)
+
+Construct an `Average` of `field` along the dimensions specified by the tuple `dims`.
+
+After the average is computed it will be stored in the `result` property.
+
+The `Average` can be used as a callable object that computes and returns the average.
+
+An `iteration_interval` or `time_interval` (or both) can be passed to indicate how often to
+run this diagnostic if it is part of `simulation.diagnostics`. `iteration_interval` is a
+number of iterations while `time_interval` is a time interval in units of `model.clock.time`.
+
+A `return_type` can be used to specify the type returned when the `Average` is
+used as a callable object. The default `return_type=Array` is useful when running a GPU
+model and you want to save the output to disk by passing it to an output writer.
+"""
+function Average(field; dims, iteration_interval=nothing, time_interval=nothing, return_type=Array)
+    dims isa Union{Int, Tuple} || @error "Average dims must be an integer or tuple!"
+    dims isa Int && (dims = tuple(dims))
```

Comment:
Now that's a feature.

Reply:
Haha `Statistics.mean` does it so I figured we should be fancy too.